Refactor `amplication-server` to use `@amplication/nest-logger-module`

ParkourKarthik opened the pull request:
<!--
Thank you for contributing to Amplication :)

PLEASE, GO THROUGH THESE STEPS BEFORE SUBMITTING A PR!

Make sure that:

1. There is an open issue for this PR. If not, please open one before submitting your changes. Before proceeding, any change needs to be discussed (You can skip this if you're fixing a typo or adding an app to the Showcase).

2. You have done your changes in a separate branch. Branches MUST have descriptive names that start with either the `fix/[issue #]-` or `feature/[issue #]-` prefixes. Good examples are: `fix/404-signin-issue` or `feature/201-new-templates`.

3. You are giving a descriptive title to your PR.

4. You are providing enough information about your changes for others to review your pull request.

-->

Issue Number: #3397

## PR Details

Refactored to use the `@amplication/nest-logger-module` in `amplication-server` instead of the `winston` logger usage.

## PR Checklist
- [x] Tests for the changes have been added
- [x] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

ParkourKarthik commented:
@souravjain540 

Need some clarification for the below:
- There is a `Logger` [used](https://github.com/amplication/amplication/pull/3553/files#diff-29bb71d326235f44c370ad50a94e373e9e2b6437dffd6392415673bae84bc2fcL9) from `@nestjs/common` apart from `winston` logger. I've retained it as the issue is to update `winston` loggers only. Let me know if I should update that as well.
- I had to remove the `RootWinstonModule` [module import](https://github.com/amplication/amplication/pull/3553/files#diff-94a399e7a5b25d2cc83b6c429b36555238df9919463184619e80f97a35f2ff54L9) along with the changes, I hope that is expected.
- There are places were `winston` is [directly used](https://github.com/amplication/amplication/pull/3553/files#diff-0bd5f3f111abe353c918c42195a3ecfc287ced63e6debf18d4d0ddda5a49297aL391) in the `build.service.ts`, I could not find a way to remove/replace the usage. Should these be retained as it is?

souravjain540 commented:
@arielweinberger please review

yuval-hazaz commented:
@arielweinberger @tupe12334 please help here

tupe12334 commented:
@ParkourKarthik Can you link us to the location of the @nest/common logger?
The removal of the RootWinstonModule is expected.
Can you please explain in which place you can't replace the Winston logger?

arielweinberger commented:
Hi, sorry for the late response and thanks for your PR.

The expectation is that the Amplication NestJS logger will be the exclusive logger module used in the server.

ParkourKarthik commented:
> @ParkourKarthik Can you link us to the location of the @nest/common logger?
> Can you please explain in which place you can't replace the Winston logger?

@tupe12334 👋 Thanks for the response. I have already mentioned the places in the comment as hyperlinks, let me know if you're having trouble in the links.

arielweinberger commented:
I see, @ParkourKarthik, thanks for pointing out. The goal for this issue is that all other forms of logging (default NestJS `Logger`, `winston` etc.) will be gone, and our package to be used exclusively.

ParkourKarthik commented:
I think I've got clarifications for other doubts except the below. Could someone help me out with the below query?

> - There are places were `winston` is [directly used](https://github.com/amplication/amplication/pull/3553/files#diff-0bd5f3f111abe353c918c42195a3ecfc287ced63e6debf18d4d0ddda5a49297aL391) in the `build.service.ts`, I could not find a way to remove/replace the usage. Should these be retained as it is?

arielweinberger commented:
@ParkourKarthik You would need to remove any usage of `winston` there and perform dependency injection to the Amplication NestJS Logger, and replace the `winston` calls there with those of our logger.

ParkourKarthik commented:
> @ParkourKarthik You would need to remove any usage of `winston` there and perform dependency injection to the Amplication NestJS Logger, and replace the `winston` calls there with those of our logger.
> 
> I suggest that you give it a try, and if you struggle, we'd love to accept your contribution anyway and tackle the `winston` instance in `build.service.ts` in a separate ticket (perhaps internally within the team).
> 
> Will you be able to resolve the conflicts? :)

Thanks, will check if I can do that and let you know. I'll rebase and fix the conflicts :+1:

ParkourKarthik commented:
@arielweinberger I've removed the all the direct references to `winston` and made some update to `nest-logger-module` and `data-service-generator` as required. I've renamed all the `winston` based variable namings as well.

souravjain540 commented:
@arielweinberger need your approval

arielweinberger commented:
@tupe12334 I see you approved this change. Have you tested it and made sure it works properly?

arielweinberger commented:
I've added some more comments. The intention is to not use Winston in any of our services, and importing Winston types into our service binds us to the Winston API. So if you could please fix that, that would be great.

Reference to Transports, etc should not be there, otherwise we are just creating an extra layer of complication "abstraction" on top of Winston without gaining much.

souravjain540 commented:
@ParkourKarthik please work on the feedback

ParkourKarthik commented on packages/amplication-nest-logger-module/src/index.ts:
```diff
@@ -1,3 +1,9 @@
 export * from "./logger.module";
 export { WINSTON_MODULE_PROVIDER as AMPLICATION_LOGGER_PROVIDER } from "nest-winston";
-export { Logger as AmplicationLogger } from "winston";
+export {
+  Logger as AmplicationLogger: Logger,
+  transports as Transports,
+  createLogger as CreateLogger,
+} from "winston";
```
The reason is because of the below question:
>There are places were winston is [directly used](https://github.com/amplication/amplication/pull/3553/files#diff-0bd5f3f111abe353c918c42195a3ecfc287ced63e6debf18d4d0ddda5a49297aL391) in the build.service.ts, I could not find a way to remove/replace the usage. Should these be retained as it is?

By the above question I meant that there were instances in `build.service` which were using `transports` and `createLogger` from `winston`.
I had to either keep it as it is or move it to `amplication-nest-logger-module`. I thought of moving as there could be similar references in other project. Let me know if I should rever these changes?

ParkourKarthik commented on packages/amplication-server/src/core/build/build.service.spec.ts:
```diff
@@ -1,10 +1,13 @@
-import { WINSTON_MODULE_PROVIDER } from 'nest-winston';
+import {
+  AMPLICATION_LOGGER_PROVIDER,
+  CreateLogger,
+  Transports
```
Will remove the test by commenting them so that it could be fixed in near future.

ParkourKarthik commented:
> I've added some more comments. The intention is to not use Winston in any of our services, and importing Winston types into our service binds us to the Winston API. So if you could please fix that, that would be great.
> 
> Reference to Transports, etc should not be there, otherwise we are just creating an extra layer of complication "abstraction" on top of Winston without gaining much.

Thanks for review. I understand the intention but I was not sure how we could replace the reference of `Transports` and `createLogger` from `winston`. That's the reason why I raised the question initially, but with lesser details on how I could solve it I had to consider doing the above changes. I could either revert the current changes on the `Transports` and `CreateLogger` references part (`winston` package will be directly referenced) or I could fix it if someone could guide me on how that could be replaced.

arielweinberger commented:
I'm happy to accept it as-is, it will still be of great value for us! Thank you @ParkourKarthik

arielweinberger commented:
@abrl91  Will test it tomorrow and hopefully we can merge it.

abrl91 commented:
Hey, @ParkourKarthik Thank you for your contribution, and congrats on your first PR.
I will take a look at your PR next week.

abrl91 commented:
run branch

yuval-hazaz commented:
@abrl91 please check and merge
